endpoints/s3/src/putobject.cpp

```diff
@@ -331,35 +434,97 @@ void irods::s3::actions::handle_putobject(
 			session_ptr->send(std::move(response));
 			return;
 		}
 		else if (upload_id.empty()) {
-			logging::error("{}: UploadPart detected but upload_id was not provided.", __FUNCTION__);
+			logging::error("{}: UploadPart detected but upload_id was not provided.", __func__);
 			response.result(beast::http::status::bad_request);
-			logging::debug("{}: returned {}", __FUNCTION__, response.reason());
+			logging::debug("{}: returned {}", __func__, response.reason());
 			session_ptr->send(std::move(response));
 			return;
 		}
 		else if (!std::regex_match(upload_id, upload_id_pattern)) {
-			logging::error("{}: Upload ID {} was not in expected format.", __FUNCTION__, upload_id);
+			logging::error("{}: Upload ID {} was not in expected format.", __func__, upload_id);
 			response.result(beast::http::status::bad_request);
-			logging::debug("{}: returned {}", __FUNCTION__, response.reason());
+			logging::debug("{}: returned {}", __func__, response.reason());
 			session_ptr->send(std::move(response));
 			return;
 		}
 
+		// see if we have enough information to stream this part directly to iRODS
+		uint64_t part_size = 0;
+		unsigned int part_number_int = 0;
+		{
+			std::lock_guard<std::mutex> guard(part_shmem::part_size_map_mutex);
+
+			// if an entry for upload id does not exist go ahead and create it
+			if (part_shmem::part_size_map.find(upload_id) == part_shmem::part_size_map.end()) {
+				part_shmem::part_size_map[upload_id] = std::unordered_map<unsigned int, uint64_t>();
+			}
+
+			try {
+				part_number_int = boost::lexical_cast<int>(part_number.c_str());
+
+				// add this part size to part_size_map
+				if (chunked_flag || special_chunked_header) {
+					// chunked - get part size from x-amz-decoded-content-length
+					// if this isn't provided then just continue without storing the part size
+					auto header = parser_message.find("x-amz-decoded-content-length");
+					if (header != parser_message.end()) {
+						part_size = boost::lexical_cast<uint64_t>(parser_message["x-amz-decoded-content-length"]);
+						part_shmem::part_size_map[upload_id][part_number_int] = part_size;
+					}
+				}
+				else {
+					// not chunked, get part size from content_length
+					part_size = boost::lexical_cast<uint64_t>(parser_message[beast::http::field::content_length]);
+					part_shmem::part_size_map[upload_id][part_number_int] = part_size;
+				}
+
+				// see if we know all of the previous part_sizes
+				know_part_offset = true;
+				for (unsigned int i = 1; i <= part_number_int - 1; ++i) {
+					if (part_shmem::part_size_map[upload_id].find(i) != part_shmem::part_size_map[upload_id].end()) {
+						part_offset += part_shmem::part_size_map[upload_id][i];
+					}
+					else {
+						know_part_offset = false;
+					}
+				}
+			}
+			catch (const boost::bad_lexical_cast&) {
+				logging::warn(
+					"{}:{} {}: part_number = {} could not parse part size as unint64_t, "
+					"continuing without storing part size in shared memory",
```

Comment:
I put a comment on this explaining the effect of continuing.

I also broke out the parsing of the part_number because we can't continue in that case.